packages/indy-vdr/src/pool/IndyVdrPoolService.ts

```diff
@@ -0,0 +1,230 @@
+    // indyLedgers config as the order of preference of ledgers
+    let value = successful.find((response) => isSelfCertifiedDid(response.value.did.nymResponse.did, response.value.did.nymResponse.verkey))?.value
+
+    if (!value) {
+      // Split between production and nonProduction ledgers. If there is at least one
+      // successful response from a production ledger, only keep production ledgers
+      // otherwise we only keep the non production ledgers.
+      const production = successful.filter((s) => s.value.pool.config.isProduction)
+      const nonProduction = successful.filter((s) => !s.value.pool.config.isProduction)
+      const productionOrNonProduction = production.length >= 1 ? production : nonProduction
+
+      // We take the first value as we take the order in the indyLedgers config as
+      // the order of preference of ledgers
+      value = productionOrNonProduction[0].value
+    }
+
+    await this.didCache.set(agentContext, did, {
+      nymResponse: {
+        did: value.did.nymResponse.did,
+        verkey: value.did.nymResponse.verkey,
+      },
+    })
+    return { pool: value.pool, did: value.did.nymResponse.did }
+  }
+
+  private async getSettledDidResponsesFromPools(did: string, pools: IndyVdrPool[]) {
+    this.logger.trace(`Retrieving did '${did}' from ${pools.length} ledgers`)
+    const didResponses = await allSettled(pools.map((pool) => this.getDidFromPool(did, pool)))
+
+    const successful = onlyFulfilled(didResponses)
+    this.logger.trace(`Retrieved ${successful.length} responses from ledgers for did '${did}'`)
+
+    const rejected = onlyRejected(didResponses)
+
+    return {
+      rejected,
+      successful,
+    }
+  }
+
+  /**
+   * Get the most appropriate pool for the given indyNamespace
+   */
+  public getPoolForNamespace(indyNamespace: string) {
+    if (this.pools.length === 0) {
+      throw new LedgerNotConfiguredError(
+        "No indy ledgers configured. Provide at least one pool configuration in the 'indyLedgers' agent configuration"
+      )
+    }
+
+    if (!indyNamespace) {
+      this.logger.warn('Not passing the indyNamespace is deprecated and will be removed in the future version.')
+      return this.pools[0]
+    }
+
+    const pool = this.pools.find((pool) => pool.config.indyNamespace === indyNamespace) // TODO check if this is corect
+
+    if (!pool) {
+      throw new LedgerNotFoundError(`No ledgers found for IndyNamespace '${indyNamespace}'.`)
+    }
+
+    return pool
+  }
+
+  private async getDidFromPool(did: string, pool: IndyVdrPool): Promise<PublicDidRequest> {
+    try {
+      this.logger.trace(`Get public did '${did}' from ledger '${pool.id}'`)
+      const request = await new GetNymRequest({ dest: did })
+
+      this.logger.trace(`Submitting get did request for did '${did}' to ledger '${pool.id}'`)
+      const response = await pool.submitReadRequest(request)
+
+      if (!response.result.data) {
+        throw new LedgerError('Not Found')
```

Comment:
We should probably also create an IndyVdrNotFoundError error based on the indy vdr error and throw that here. Aslo can you make the error message a bit more descriptive?